cloudinit/sources/__init__.py

```diff
@@ -392,6 +395,11 @@ def get_vendordata(self):
             self.vendordata = self.ud_proc.process(self.get_vendordata_raw())
         return self.vendordata
 
+    def get_vendordata2(self):
```

Comment:
this should not have been put here.  A single datasource (openstack) has a thing called 'vendordata2'.  Shouldn't the change have been made in openstack datasource rather than propagating up to the base class?

Reply:
Hello @smoser. I'm looking at how to move this handling into openstack-specific code and it's starting to seem like a pretty big refactor.  It's easy enough to move the things from __init__.py into DataSourceOpenStack.py but then the vendordata2 handlers in stages.py will need to be made aware of which subclasses do or don't support vendordata2; failing that we could subclass things in stages.py but as best I can tell that isn't currently done anywhere so would require some additional scaffolding.

If you still want me to go ahead with refactoring I can work on it but I welcome your thoughts about how to avoid littering the existing code with checks :)  It might also be worth creating a new bug to track.

(Of course when I wrote the initial patchset I was imagining that this might be a feature that other datasources would want to implement but that may not be realistic.)

-Andrew

Reply:
> (Of course when I wrote the initial patchset I was imagining that this might be a feature that other datasources would want to implement but that may not be realistic.)

I'm confused as to how that would be any different than `vendor data` to cloud-init.

As far as rework, i would have thought that you would just change the datasource to say : If there is vendor-data2, present it as vendor-data, otherwise use vendor-data.